spec/models/sms/sms_decoder_spec.rb

```diff
@@ -531,6 +531,83 @@
     end
   end
 
+  describe "nested groups" do
+    let(:form) do
+      create_form(
+        questions: %w[integer multilevel_select_one_as_text_for_sms],
+        default_option_names: true
+      )
+    end
+
+    # QingGroup (root)
+    #   Questioning
+    #   Questioning (multilevel)
+    #   QingGroup
+    #     Questioning
+    #     Questioning
+    # TODO: add another nested group
+    #     QingGroup
+    #       Questioning
+    before do
+      qing_group = create_form(questions: %w[integer integer]).root_group
+      qing_group.parent = form.root_group
+      qing_group.rank = 3
+      qing_group.save!
+    end
+
+    # 1. AnswerGroup
+    #   1. Answer
+    #   2. AnswerSet
+    #     1. Answer
+    #     2. Answer
+    #   3. AnswerGroup
+    #     1. Answer
+    #     2. Answer
+    it "builds corresponding answer hierarchy" do
+      response = create_response(body: "#{form.code} 1.1 2.tulip 3.3 4.4")
+      response.answer_hierarchy.try(:save, response)
+
+      answer_group = response.root_node
+      expect(answer_group).to be_a(AnswerGroup)
+      expect(answer_group.new_rank).to eq 1
+      expect(answer_group.children.count).to eq 3
+
+      answer = answer_group.children[0]
+      expect(answer).to be_a(Answer)
+      expect(answer.value).to eq "1"
+      expect(answer.new_rank).to eq 1
+
+      answer_set = answer_group.children[1]
+      expect(answer_set).to be_a(AnswerSet)
+      expect(answer_set.new_rank).to eq 2
+      expect(answer_set.children.count).to eq 2
+
+      answer = answer_set.children[0]
+      expect(answer).to be_a(Answer)
+      expect(answer.new_rank).to eq 1
+      expect(answer.option.name_en).to eq "Plant"
+
+      answer = answer_set.children[1]
+      expect(answer).to be_a(Answer)
+      expect(answer.new_rank).to eq 2
+      expect(answer.option.name_en).to eq "Tulip"
+
+      answer_group = answer_group.children[2]
+      expect(answer_group).to be_a(AnswerGroup)
+      expect(answer_group.new_rank).to eq 3
+
+      answer = answer_group.children[0]
+      expect(answer).to be_a(Answer)
+      expect(answer.new_rank).to eq 1
+      expect(answer.value).to eq "3"
+
+      answer = answer_group.children[1]
+      expect(answer).to be_a(Answer)
+      expect(answer.new_rank).to eq 2
+      expect(answer.value).to eq "4"
```

Comment:
Please use/modify the `expect_children` method from `spec/models/results/response_tree_builder_spec.rb` for a more compact way of checking the hierarchy. @DevneyHamilton is also using this method and has moved it into a separate file (in her branch). Devney I wonder if you could share that file or diff with Scott and he could include it so the merge process will be easier?

Reply:
Looking at this further -- you both will probably be needing to make some changes in order to be able to check for appropriate values in the hierarchy (the existing `expect_children` method doesn't check values, only ranks, qing_ids and types). I think it would be great if you could work together on this.